compiler/src/dotty/tools/dotc/core/classfile/ClassfileParser.scala

```diff
@@ -518,8 +518,8 @@ class ClassfileParser(
         else {
           val elems = arr.toList
           val elemType =
-            if (elems.isEmpty) defn.ObjectType
-            else ctx.typeComparer.lub(elems.tpes).widen
+            if (elems.isEmpty) WildcardType // No way to figure out the element type without forcing the annotation constructor
+            else elems.head.tpe.widen
```

Comment:
What does Scala'2 ClassfileParser do? As far as I remember, this could came directly from it. 

I am unsure whether we can simply take `head`. What about annotation arguments of different types? Are these not possible? An alternative would be to use
```
ctx.typeComparer.lub(elems.tpes).widenUnion
```
That seems to be the safer fix.

Reply:
> What does Scala'2 ClassfileParser do? As far as I remember, this could came directly from it.

In Scala 2, it seems that arguments to annotations are represented using a special data structure which is not a tree. In particular, arrays are represented as `ArrayAnnotArg`, I assume we don't want to bring all the complexity associated with this back.

> I am unsure whether we can simply take head. What about annotation arguments of different types? Are these not possible?

I went through the Java and JVM spec and as far as I can tell there's only one case which is slightly problematic, it involves nested arrays:
```scala
values = { {}, { 1 } }
```
in that case, if we only look at the type of the first element, we'll type the full argument as `Array[Array[<?>]]` instead of `Array[Array[Int]]`, this still typechecks but is less precise.

> An alternative would be to use `ctx.typeComparer.lub(elems.tpes).widenUnion`

This would probably work too, but I think avoiding `lub` would be better:  just today, someone reported on gitter a cyclic reference in their project, which I traced to this precise `lub` call which indirectly calls `ctx.scala2CompatMode`, switching to `elems.head.tpe.widen` fixed the cycle. (An alternative solution would be to make `lub` behave differently when unpickling a Java unit).

Overall, I think the choice made by this PR is the best solution for now. Any other solution will require some extra complexity in the compiler and so should be motivated by an actual usecase.